File: 2_semester/Laba_5/task_2.cpp
#include <map>
#include <set>
#include <string>
#include <iostream>

void add_friend(std::map <std::string, std::set<std::string>>& friends, std::string name_1, std::string name_2)
{
	friends[name_1].insert(name_2);
	friends[name_2].insert(name_1);
}
int count_friend(std::map <std::string, std::set<std::string>> friends, std::string name_1)
{
	return friends[name_1].size();
}
bool is_friend(std::map <std::string, std::set<std::string>> friends, std::string name_1, std::string name_2)
{
	for (auto item : friends[name_1])
		if (item == name_2)
			return true;
	return false;
}

int main()
{
	std::map <std::string, std::set<std::string>> friends;
	while (true) {
		std::string comand, name_1, name_2;
		std::cin >> comand >> name_1;

		if (comand != "COUNT" && comand != "1")
			std::cin >> name_2;

		if (comand == "FRIENDS" || comand == "0")
			add_friend(friends, name_1, name_2);
		if (comand == "COUNT" || comand == "1")
			std::cout << count_friend(friends, name_1) << std::endl;
		if (comand == "QUESTION" || comand == "2") {
			is_friend(friends, name_1, name_2) ? std::cout << "YES" : std::cout << "NO";
			std::cout << std::endl;
		}

	}
}

File: 2_semester/Laba_5/task_2.py

def add_friend(friends, name_1, name_2):
	if name_1 == name_2:
		return
	if type(friends.get(name_1)) != type(set()):
		friends[name_1] = set()
	if type(friends.get(name_2)) != type(set()):
		friends[name_2] = set()
	friends[name_1].add(name_2)
	friends[name_2].add(name_1)

def count_friend(friends, name_1):
	if type(friends.get(name_1)) != type(set()):
		return 0
	return len(friends.get(name_1))

def is_friend(friends, name_1, name_2):
	if type(friends.get(name_1)) != type(set()):
		return False
	for item in friends.get(name_1):
		if item == name_2:
			return True
	return False



friends = {'a': {'b'}, 'b' : {'a'}}
while True:
	args = list(input().split())
	comand = args[0]
	if (comand == "FRIENDS" or comand == "0"):
		add_friend(friends, args[1], args[2])
	if (comand == "COUNT" or comand == "1"):
		print(count_friend(friends, args[1]))
	if (comand == "QUESTION" or comand == "2"):
		if is_friend(friends, args[1], args[2]):
			print('YES')
		else:
			print('NO')